Problem (capa) feedback rework

alisan617 opened the pull request:
This is now the feature branch for the [Problem Feedback Epic](https://openedx.atlassian.net/browse/TNL-4877). Individual stories merged are represented as commits.

[sandbox](https://submit-capa.sandbox.edx.org)

## Testing
- [x] a11y
- [x] i18n
- [x] RTL
- [x] safecommit process
- [x] Unit, integration, acceptance tests as appropriate

## Reviewers
- [ ] @cahrens 
- [x] @staubina 
- [ ] @bjacobel
- [ ] @edx/edx-accessibility
- [ ] UX @chris-mike 
- [ ] Product @sstack22 
- [ ] Doc @catong 

## Post-review
- [x] Squash commits

## Visual update
<img width="773" alt="before" src="https://cloud.githubusercontent.com/assets/16127998/17228383/bd794458-54e0-11e6-9683-156b585a92f0.png">

<img width="852" alt="after" src="https://cloud.githubusercontent.com/assets/16127998/17228384/bd79a48e-54e0-11e6-99e5-6de68ca10102.png">

cahrens commented:
@alisan617 Can you provide a link to a unit where the problems have all the button types enabled (on your sandbox)?

"Show Answer" feels like it is not vertically aligned properly. It feels a bit "too low" compared to the Submit button.

![image](https://cloud.githubusercontent.com/assets/484484/17064989/8645da14-500d-11e6-9187-ad6d2e18f18e.png)

cahrens commented on common/lib/xmodule/xmodule/capa_base.py:
```diff
@@ -413,8 +413,7 @@ def check_button_name(self):
+        check = _('Submit')
 
         # Apply customizations if present
         if 'custom_check' in self.text_customization:
```
This customization code (which is now dead) can be removed in https://openedx.atlassian.net/browse/TNL-4939.

Note that the documentation block for check_button_name also needs to be updated. I suppose that can also be done in TNL-4939.

cahrens replied:
I have create a PR to remove the text_customization code-- #13080.

cahrens commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -432,7 +432,7 @@ div.problem {
 
     > span {
-      display: block;
+      //display: block;
```
Don't forget to delete this commented out line.

cahrens commented on common/lib/xmodule/xmodule/tests/test_capa_module.py:
```diff
@@ -991,76 +991,75 @@ def test_save_problem_submitted_no_randomize(self, rerandomize):
-        self.assertEqual(module.check_button_checking_name(), "Checking...")
+        self.assertEqual(module.check_button_checking_name(), "Submitting...")
 
     def test_check_button_name_customization(self):
```
This text customization is no longer working, so I would advocate just deleting the tests (all of the associated code will be deleted as part of https://openedx.atlassian.net/browse/TNL-4939). "custom_check" and "custom_final_check" are no longer used at all in capa_base.py, so all this test is actually doing now is ensuring what the hard-coded values of the button are.

alisan617 replied:
I cherry-picked your PR, this part should be gone.

cahrens commented on lms/templates/problem.html:
```diff
@@ -18,25 +18,37 @@ <h3 class="hd hd-2 problem-header">
+    % endif
+    </div>
+    % if check_button:
+    <button class="check ${ check_button } btn-brand" data-checking="${ check_button_checking }" data-value="${ check_button }">
```
Do you understand how data-checking is used? I can't figure out why it is needed, and am hoping we can just rip it out. Perhaps it is vestigial from all of the "custom button name" code that can also be removed.

I'll put a note in TNL-4939 about it.

alisan617 replied:
There is disabled/enabled state applied to the Submit button, when it is "submitting" the button is disabled. The attribute is used to determine which state the button is in.

See line 34 in caps/display.coffee

cahrens replied:
From my read of the code, it looks like the data attribute is only used to get what the name should be during the time that the "checking" is in progress. So it's fine to leave as-is if we wish the names to be sent from the server (instead of hard-coding the names in the JS or HTML).

cahrens commented on common/lib/xmodule/xmodule/tests/test_capa_module.py:
```diff
@@ -991,76 +991,75 @@ def test_save_problem_submitted_no_randomize(self, rerandomize):
 
     def test_check_button_name(self):
```
I'd say this test case can be simplified quite a bit since the button never change its name anymore to "Final ...".

alisan617 commented:
jenkins run quality

alisan617 commented:
jenkins run bokchoy

cahrens commented on lms/djangoapps/courseware/features/problems.py:
```diff
@@ -151,7 +151,7 @@ def action_button_present(_step, buttonname, doesnt_appear):
     # The label text is changed by static/xmodule_js/src/capa/display.js
     # so give it some time to change on the page.
-    label_css = 'button.show span.show-label'
+    label_css = 'button.show.btn.btn-default.btn-small span.show-label'
```
Why is it necessary to specify all the additional classes (btn.btn-default.btn-small)?

alisan617 replied:
btn can be removed, but btn-small is needed, as design spec has smaller font for it

ahsan-ul-haq commented:
@alisan617 I think the issue with your failing bokchoy test is that test calls `complete_course_problems` [here](https://github.com/edx/edx-platform/blob/master/common/test/acceptance/tests/lms/test_certificate_web_view.py#L188) and it expect css of check button to have class Check [here](https://github.com/edx/edx-platform/blob/master/common/test/acceptance/tests/lms/test_certificate_web_view.py#L228) which seemingly you have changed.

alisan617 commented:
@ahsan-ul-haq  that solved the problem. All tests passed. Thank you.

cahrens commented on common/test/acceptance/tests/lms/test_certificate_web_view.py:
```diff
@@ -225,7 +225,7 @@ def complete_course_problems(self):
 
         # Submit the answer
-        self.course_nav.q(css='button.check.Check').click()
+        self.course_nav.q(css='button.check.Submit').click()
```
Ah yes-- I didn't think that the CSS classes had changed. But I should have had you run the test locally, which might have helped to isolate the issue.

But @alisan617, why is there a css class of "Submit" at all? If the "Submit" button is the only one with the class "check", I recommend dropping the "Submit" class entirely, as it is not our style to have CSS classes that match (English) UI text. I would think that "button.check" is sufficient.

cahrens commented on common/lib/xmodule/xmodule/js/fixtures/problem_content.html:
```diff
@@ -12,11 +12,13 @@ <h3 class="hd hd-2 problem-header">Problem Header</h3>
     <span id="input_example_1_dynamath"></span>
 
-    <button class="check Check" data-checking="Checking..." data-value="Check"><span class="check-label">Check</span><span class="sr"> your answer</span></button>
+    <button class="check Submit" data-checking="Submitting..." data-value="Submit"><span class="check-label">Submit</span><span class="sr"> your answer</span></button>
```
Update this fixture file to match the production one ("Submit" class no longer present).

staubina replied:
Nit: Should the class be capitalized?  I am still learning the style guidelines

Do we need translations on this text?

cahrens commented on common/lib/xmodule/xmodule/capa_base.py:
```diff
@@ -408,44 +402,24 @@ def check_button_name(self):
-        Return the "checking..." text for the "check" button.
+        Return the "submitting..." text for the "submit" button.
 
         After the user presses the "check" button, the button will briefly
```
Might as well update this documentation as well.

cahrens commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -906,47 +905,79 @@ div.problem {
+    }
+    &.btn-brand {
+      &:hover {
+        background-color: #0ea6ec;
```
Are these color values (here and below) from the pattern library? If so, it would be better to define variables that reference the pattern library, similar to https://github.com/edx/edx-platform/pull/12950/files#diff-8e9589dd2238bbdf2a4fca2969a3e963R44.

staubina replied:
I am going to second the use of variables, especially if they are used throughout the css, it makes it easier to expand and add on later.

cahrens commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -906,47 +905,79 @@ div.problem {
+        background-color: #0ea6ec;
+      }
+      &.check {
+        &.is-disabled, &.disabled {
```
Why both is-disabled and disabled?

bjacobel replied:
Nit: We prefer the CSS style of one class per line, so this would be:

```css
&.check {
  &.is-disabled,
  &.disabled {
    ...
  }
}
```

cahrens commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -906,47 +905,79 @@ div.problem {
 
-  .save {
-    @extend .blue-button !optional;
+  @media (min-width: 768px) {
```
What happens when width < 768px? How did you pick that value?

staubina replied:
Are we concerned about adjusting for mobile screens?

alisan617 replied:
Per UX, yes. The buttons to the right have the tab index before the Submit button. For smaller screen, they are on top of the Submit button.

cahrens replied:
@alisan617 and @chris-mike I understand moving the buttons above Submit for smaller screens. But it looks very cramped and inelegant-- we need to make sure this will look good on mobile as well. Is there a way to do better padding in this case?

![image](https://cloud.githubusercontent.com/assets/484484/17234704/4d85df52-5509-11e6-9698-e58dcece7141.png)

cahrens commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -906,47 +905,79 @@ div.problem {
-    .show-label {
-      font-weight: 600;
-      font-size: 1.0em;
+  .save, .show, .reset, .hint-button {
```
It might be cleaner to add a common class to all these buttons.

cahrens commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -906,47 +905,79 @@ div.problem {
+    font-size: 0.9em;;
     -webkit-font-smoothing: antialiased;
+
+    @media (min-width: 768px) {
```
Can you explain this?

In general, after @chris-mike has had a chance to review this PR I would like another FED (perhaps Brian?) to review it. I'm not very strong in CSS, and I don't know what all of our best practices are.

cahrens replied:
@bjacobel Would you have time to review the CSS later today or tomorrow?

bjacobel replied:
Yes - I don't know the background of this ticket (what it fixes/how) but I can review the CSS for style.

cahrens commented on common/lib/xmodule/xmodule/tests/test_capa_module.py:
```diff
@@ -991,76 +991,35 @@ def test_save_problem_submitted_no_randomize(self, rerandomize):
+        self.assertEqual(module.check_button_name(), "Submit")
 
         module = CapaFactory.create(attempts=attempts, max_attempts=attempts)
-        self.assertEqual(module.check_button_name(), "Final Check")
```
As mentioned above, I'd recommend simplifying this test point considerably since there is no more "Final Check", and the text on the button is always the same.

cahrens commented on lms/templates/problem.html:
```diff
@@ -18,25 +18,37 @@ <h3 class="hd hd-2 problem-header">
-    % if reset_button:
-    <button class="reset" data-value="${_('Reset')}">${_('Reset')}<span class="sr"> ${_("your answer")}</span></button>
+    <span class="problem-action-button-wrapper">
+        <button class="hint-button btn btn-default btn-small" data-value="${_('Hint')}" title="${_('Hint')}"><span class="icon fa fa-question" aria-hidden="true"></span>${_('Hint')}</button>
```
Why does there need to be a title given that the button has visible text? Forgive me if I've asked this previously.

alisan617 replied:
will remove

cahrens commented:
@alisan617 This is looking good. Please make sure the sandbox is updated after responding to my review feedback, and let's try to get reviews done tomorrow so that we might be able to close this story by Thursday afternoon.

staubina commented:
@alisan617 I might be looking in the wrong place, but I should see the new buttons on the Demo Exam, right?

staubina commented on common/lib/xmodule/xmodule/capa_base.py:
```diff
@@ -408,44 +402,24 @@ def check_button_name(self):
+
+        return check
 
     def check_button_checking_name(self):
```
I would consider adjusting the method name to something more generic, not that we are no longer using "checking"

cahrens replied:
I put some notes about cleanup like this in https://openedx.atlassian.net/browse/TNL-4939.

staubina commented on common/lib/xmodule/xmodule/capa_base.py:
```diff
@@ -408,44 +402,24 @@ def check_button_name(self):
-
         _ = self.runtime.service(self, "i18n").ugettext
-        return _('Checking...')
+        return _('Submitting...')
```
@cahrens  Will the ... mess with the translation recognition?  Should we leave just "Submitting" in the translation portion?

cahrens replied:
@staubina Actually, you don't want to concatenate "..." onto the end of a translated string, as it does not make sense for all languages (so what is here is better).

But I think we've actually moved away from putting "..." on strings in general. @catong should the ... be dropped?

staubina replied:
Ok, that makes sense.

catong replied:
@cahrens @staubina Is this string displaying on the button, or is it plain text in the body of the problem that replaces the button? Do you have a screenshot you can share?

alisan617 replied:
@catong I have my sandbox up and running https://alisan617.sandbox.edx.org

cahrens replied:
It displays on the button during the time that the submission is in process.

catong replied:
Thanks @alisan617! I just looked at the button in your sandbox. I agree with @cahrens' recommendation to get rid of the ellipses.

staubina replied:
Has the Sanbox been updated with the removal of the elipses?  I am still seeing it on the submitting button.

staubina commented on common/lib/xmodule/xmodule/tests/test_capa_module.py:
```diff
@@ -991,76 +991,35 @@ def test_save_problem_submitted_no_randomize(self, rerandomize):
         attempts = random.randint(1, 10)
         module = CapaFactory.create(attempts=attempts - 1, max_attempts=attempts)
-        self.assertEqual(module.check_button_name(), "Final Check")
+        self.assertEqual(module.check_button_name(), "Submit")
```
If you are reusing "Submit" so much, should it be a variable?

cahrens replied:
I think this test can be boiled down to a single check now, since there is no more "Final Check", and the text on the button is always the same.

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -906,47 +905,79 @@ div.problem {
 // ====================
 div.problem .action {
-  margin-top: $baseline;
+  @include margin($baseline 0 $baseline 0);
```
nit: I believe `($baseline 0 $baseline 0)` can be expressed as just `($baseline 0)` (although I don't know if the internal implementation of the `margin()` sass mixin plays by the same rules as plain css `margin` prop is so this could be wrong)

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -906,47 +905,79 @@ div.problem {
+        &.is-disabled, &.disabled {
+          background: transparent;
+          border-color: #d2d0d0;
+          color: #6b6969;
```
Couple more colors here that should be variables, if possible.

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -906,47 +905,79 @@ div.problem {
+
+    .problem-action-buttons-wrapper {
+      position: absolute;
+      @include margin-right(($baseline*.75));
```
Nit: Mixins/extends should be the first lines in a Sass block - first any `%extends` / `@include` / other Sass things, then normal CSS props, then any nested blocks.

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -906,47 +905,79 @@ div.problem {
+    border-right: 1px solid #ddd;
+    &:last-child {
+      border: none;
+    }
```
This is a good example of how to order the different kinds of things I was talking about above!

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -906,47 +905,79 @@ div.problem {
-    margin-top: 8px;
     color: $gray-d1;
-    font-style: italic;
+    font-size: 0.9em;;
```
Like the colors, font sizes should also should use common variables if possible.

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -906,47 +905,79 @@ div.problem {
     -webkit-font-smoothing: antialiased;
+
+    @media (min-width: 768px) {
+      font-size: 0.8em;
```
Consider using `rem`s instead of explicitly changing font size based on screen width.

(No idea if that's actually feasible given the existing CSS/markup in this area of the codebase, but I thought I'd mention it.)

bjacobel replied:
Oh, also,  just realized you're shimming the Pattern Library, so you should be using the [Pattern Library font sizes](http://ux.edx.org/design_elements/copy/) if you can - eg. `@extend %copy-micro;`

alisan617 replied:
can't use rem yet, the font becomes too small in Studio

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -1505,4 +1536,4 @@ div.problem .annotation-input {
     background: url('#{$static-path}/images/partially-correct-icon.png') center center no-repeat;
   }
-}
+}
```
Looks like the newline at the end of the file was removed - all files should end in a newline.

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -906,47 +905,79 @@ div.problem {
 }
 
+
 // +Problem - Misc, Unclassified Mess Part 2
```
lol. [this is fine](https://gifs.bjacobel.com/thisisfine.gif)

alisan617 replied:
This wasn't from me. I had the same reaction.

bjacobel commented on common/static/sass/edx-pattern-library-shims/_buttons.scss:
```diff
@@ -0,0 +1,121 @@
+// ------------------------------
```
Not sure what's up with this, could you explain? Looks like a bunch of the pattern library styles just copied in to a new folder? We should just be using the real pattern library directly, if possible - it's installed in Platform already.

alisan617 replied:
This is css v1. We have UXPL in v2. We can't use v2 here, the layout will break. That's why this comes the shim folder in the v1.

bjacobel replied:
@andy-armstrong, ideas about what to do here? Is this the right way to go about getting UXPL stuff in v1 CSS?

andy-armstrong replied:
This approach was being used before, and I don't have a better idea in mind, so it works for me.

bradenmacdonald replied:
@e-kolpakov ^ Checkout this discussion and this PR - may provide a simpler approach for https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/86 though perhaps it's too late.

e-kolpakov replied:
@bradenmacdonald Well, it will be useful when it lands; until than we can't use it. And after it lands it's just removing 5-6 styles from DnDv2 own css anyway.

Oh, and I don't think we want to convert DnDv2 css to sass/scss, at least as part of  edx-solutions/xblock-drag-and-drop-v2#86

chris-mike commented:
Christina notes above that the alignment of the "show answer" "buttons" seem off. @sstack22  has similar concerns for alignment also being off... concerns are for vertical alignment, primarily, but I think @cahrens comment above references horizontal alignment.  I had an offline convo with @alisan617  and @sstack22 about addressing this and some other issues that have come up as it relates to responsiveness.  These will be addressed in a different PR.

catong commented:
@alisan617 I am going to tag @lamagnifica for doc review of this PR as she has been more involved than I have, with the problem-level a11y changes.

lamagnifica commented:
this looks good to me (thanks for the sandbox!) Please tag me if any questions about labels come up.  👍

alisan617 commented:
@chris-mike I have to pick another color for disabled button font. The colors you provided didn't meet the contrast ratio of 4.5.

alisan617 commented:
ready for re-review!

clrux commented:
@alisan617 If the button is `disabled`, then it doesn't need to pass contrast checks. The [W3C states](https://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-contrast.html) that:

```
1.4.3 Contrast (Minimum): The visual presentation of text and images of text has a contrast ratio of at least 4.5:1, except for the following: (Level AA)

- Large Text: Large-scale text and images of large-scale text have a contrast ratio of at least 3:1;
-Incidental: Text or images of text that are part of an inactive user interface component, that are pure decoration, that are not visible to anyone, or that are part of a picture that contains significant other visual content, have no contrast requirement.
- Logotypes: Text that is part of a logo or brand name has no minimum contrast requirement.
```

The second bullet point here is the key. I'd use what @chris-mike supplied.

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -906,47 +907,76 @@ div.problem {
-  .show {
+  .problem-action-button-wrapper {
+    @include padding(($baseline/10) 0 $baseline ($baseline/4));
+    border-right: 1px solid #ddd;
```
Some colors that aren't variables still around here.

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -906,47 +907,76 @@ div.problem {
-    margin-top: 8px;
     color: $gray-d1;
-    font-style: italic;
+    font-size: 0.8em;
```
Font size still hard-coded, should be a variable.

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -906,47 +907,76 @@ div.problem {
     -webkit-font-smoothing: antialiased;
+
+    @media (min-width: 768px) {
+      font-size: 0.9em;
```
Ditto here.

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -906,47 +907,76 @@ div.problem {
+
+    .problem-action-buttons-wrapper {
+      position: absolute;
+      right: $baseline*1.5;
```
Should space around operators like we do in js - eg `$baseline * 1.5`

alisan617 replied:
not a thing exists in all scss files now. This could be a nitpick/linting feature in future. some have (), some don't.

bjacobel replied:
We don't currently have a styleguide for sass, but the [sass docs use spaces around operators](http://sass-lang.com/guide#Operators), so IMO that's what we should be doing.

alisan617 commented:
@clrux @chris-mike I reverted the colors as you asked. The color contrast failure is back despite the button is disabled. Could you provide guideline how to fix this?

alisan617 commented:
@staubina sandbox is updated

alisan617 commented:
Ready for re-review. Sandbox is updated.

cahrens commented:
@alisan617 Can you provide a link to a unit where the problems have all the button types enabled (on your sandbox)?

alisan617 commented:
@cahrens  here you go. **2nd question down**
https://alisan617.sandbox.edx.org/courses/course-v1:edX+DemoX+Demo_Course/courseware/1414ffd5143b4b508f739b563ab468b7/workflow/4?activate_block_id=block-v1%3AedX%2BDemoX%2BDemo_Course%2Btype%40vertical%2Bblock%40f91d8d31f7cf48ce990f8d8745ae4cfa

bjacobel commented:
LGTM.

cahrens commented:
@chris-mike Currently (in production) when you have used up all of the attempts for a problem, the Check/Final Check button disappears. This looks pretty odd now-- what behavior do you want? We can clean it up in https://openedx.atlassian.net/browse/TNL-4939.

![image](https://cloud.githubusercontent.com/assets/484484/17234577/3b89f1c2-5508-11e6-8d71-e7650e416988.png)

cahrens commented:
@alisan617 My only remaining concern is how the layout looks for narrow displays (see my comment above with screenshot). Otherwise 👍 -- good work!

staubina commented:
:+1: Looks good to me.

alisan617 commented:
jenkins run a11y

cahrens commented:
@alisan617 your change looks good to me (though as I mentioned, we need to remember to check all this work on a phone at some point). 👍 

@edx/edx-accessibility @sstack22 @chris-mike will you be giving another look at this PR, or do you plan to review at a later stage?

sstack22 commented:
@cahrens @alisan617 - I'm sorry for the delay in my review. This looks great. Only concerns is that when you minimize the screen to a certain extent, the submissions text and hint button overlap. Is it possible for the "You have used 2 of 20 submissions" to be moved beneath the submit button in this situation?

![submit pr](https://cloud.githubusercontent.com/assets/11049466/17373681/5016a5b8-5977-11e6-9c66-b832d4e215d8.png)

I've also noticed the wireframes change the text from "submissions" to "attempts." I believe this text change is covered by TNL-4939, but wanted to mention, as it also gives more space on that line.

cahrens commented:
That's funny-- I just noticed that issue in https://github.com/edx-solutions/xblock-drag-and-drop-v2/pull/86, but hadn't noticed it was here as well. A fix should be made so that the Hint/Save/Show Answer buttons should wrap up above Submit before any overlapping happens.

alisan617 commented:
@sstack22 I noticed that overlapping issue, it is a "to-do" in 4939. However, I didn't know of "attempts" replacing "submissions"

cahrens commented:
I will make a note of both things in TNL-4939, as I don't believe they were explicitly called out before.

alisan617 commented:
jenkins run lettuce

andy-armstrong commented on common/static/sass/edx-pattern-library-shims/_buttons.scss:
```diff
@@ -0,0 +1,121 @@
+// LMS Problem Feedback Revamp styling
+// Mirror styles from the Pattern Library
+
+// NOTE: This is the left-to-right (LTR) configured style compile.
```
I would remove this comment, and make this file work for both directions. There isn't an RTL version...

andy-armstrong commented on common/static/sass/edx-pattern-library-shims/_buttons.scss:
```diff
@@ -0,0 +1,121 @@
+
+    &:only-child,
+    .sr-only + & {
+      margin-right: 0;
```
This should be as follows to be bi-directional:

    @include margin-right(0);

andy-armstrong commented on common/static/sass/edx-pattern-library-shims/_buttons.scss:
```diff
@@ -0,0 +1,121 @@
+    background: $btn-brand-disabled-background;
+    color: $btn-brand-disabled-color;
+  }
+}
```
Nit: missing newline.

bjacobel replied:
+1

andy-armstrong commented on cms/static/sass/_build-v1.scss:
```diff
@@ -87,3 +87,6 @@
 @import 'elements/codemirror-overrides';
+
+// CAPA Problem Feedback
+@import 'edx-pattern-library-shims/buttons';
```
Nit: missing newline.

andy-armstrong commented:
@alisan617 This looks fantastic. When will it merge to master, as I'd like to use the shims for discussions. If not soon, could we refactor those changes out into a separate PR on master. I think a lot of teams could benefit from having the shims available.

alisan617 commented:
jenkins run lettuce

cahrens commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -153,6 +153,14 @@ div.problem {
     }
   }
 
+  span > label {
```
It looks like this commit should be merged with commit https://github.com/edx/edx-platform/pull/13052/commits/c4fbcddbbc88f789ad7bc32c1b997e94bbc80c15 to get rid of these changes (they all seem to be reverted in that commit).

cahrens commented on common/lib/capa/capa/tests/response_xml_factory.py:
```diff
@@ -72,10 +72,6 @@ def build_xml(self, **kwargs):
             script_element.set("type", "loncapa/python")
             script_element.text = str(script)
 
-        # The problem has a child <p> with question text
```
Moved question text into proper location-- within response type and using a label.

cahrens commented on common/lib/capa/capa/tests/response_xml_factory.py:
```diff
@@ -113,9 +113,13 @@ def textline_input_xml(**kwargs):
         """
         math_display = kwargs.get('math_display', False)
         size = kwargs.get('size', None)
+        input_element_label = kwargs.get('input_element_label', '')
```
For the problem that has multiple text input fields, the question text is for the whole problem, but each text input also needs its own label. This is consistent with our example template.

cahrens commented on common/lib/capa/capa/tests/response_xml_factory.py:
```diff
@@ -267,9 +271,6 @@ def create_response_element(self, **kwargs):
         *answer_attr*: The "answer" attribute on the tag itself (treated as an
         alias to "expect", though "expect" takes priority if both are given)
-
-        *group_label*: Text to represent group of inputs when there are
```
Group label had been added in for the problem with multiple text inputs, but once question text was moved into the correct location, group label was no longer needed.

cahrens replied:
Actually, I need to follow up with @muhammad-ammar and/or @muzaffaryousaf about this. I saw in the code that "group_label" does exist in code, so I need to understand the interaction between it and question text. This test might need to be changed again to restore coverage of group_label.

muzaffaryousaf replied:
@cahrens The group label is like main question text and then individual "inputs" have their own label attribute. You can check detailed version here https://openedx.atlassian.net/wiki/display/A11Y/Custom+Python+Evaluated+Input 
Let me know if you still have any question(s).

cahrens commented on common/lib/capa/capa/tests/test_html_render.py:
```diff
@@ -151,10 +151,6 @@ def test_render_response_xml(self):
         self.assertEqual(rendered_html.tag, "div")
 
-        # Expect question text is in a <p> child
-        question_element = rendered_html.find("p")
```
The question text is no longer present in the rendered html because it is added in the response element. And that callback is mocked, so all that we can check is that it is passed in the context (see line 184).

cahrens commented on common/test/acceptance/tests/lms/test_certificate_web_view.py:
```diff
@@ -218,11 +218,11 @@ def complete_course_problems(self):
 
         # Select correct radio button for the answer
-        self.course_nav.q(css='fieldset div.field:nth-child(3) input').nth(0).click()
+        self.course_nav.q(css='fieldset div.field:nth-child(4) input').nth(0).click()
```
These changes were necessary because multiple_choice.xml was fixed (right above).

cahrens commented on common/test/acceptance/tests/lms/test_lms_user_preview.py:
```diff
@@ -234,7 +234,6 @@ def test_student_state_delete_for_problem_loaded_via_ajax(self):
                          'for user {}'.format(self.USERNAME), msg)
 
 
-@attr(shard=3)
```
Moving this to individual test points so that the a11y test point included in this class does not also run in the bok choy phase.

cahrens commented on common/test/acceptance/tests/lms/test_problem_types.py:
```diff
@@ -286,9 +286,6 @@ def test_problem_type_a11y(self):
-                'aria-allowed-attr',  # TODO: AC-491
-                'aria-valid-attr',  # TODO: AC-491
-                'aria-roles',  # TODO: AC-491
                 'checkboxgroup',  # TODO: AC-491
```
Removing these last 3 ignores introduce test failures. Checkboxgroup and radiogroup ignores are related to the funkier radio/checkbox + text input field problems. Removing section introduces a number of different types of failures-- FYI @clrux, as AC-491 has been closed.

cahrens replied:
Created https://openedx.atlassian.net/browse/TNL-5576 for checkboxgroup and radiogroup issues.

cahrens commented on common/lib/capa/capa/templates/formulaequationinput.html:
```diff
@@ -19,9 +19,7 @@
-            <span class="sr">
-                ${status.display_name}
-            </span>
+            <span class="sr">${status.display_tooltip}</span>
```
The Lahore PR introduced these changes, and @alisan617 added the parent tag aria-describedby attributes in her PR. While resolving merge conflicts, the Lahore changes got lost (which thankfully had unit test coverage!).

cahrens commented on common/test/acceptance/pages/lms/problem.py:
```diff
@@ -108,32 +110,80 @@ def fill_answer_numerical(self, text):
+
+    def is_save_button_enabled(self):
+        """ Check for the visibility of the Save button """
+        return not self.q(css='.action .save').attrs('disabled') is None
```
@staubina This method isn't doing what you think it is (I discovered this when basing another method off of this code).

When the button has no "disabled" attributes, self.q(css='.action .save').attrs('disabled') returns a list [None]. That isn't None, but it has nothing to do with whether or not the button is disabled.... When the button IS disabled, you get an array with other values ('true', I believe).

My point is that this method will always return True, regardless of the disabled state. Can you please fix it and investigate the other method(s) that check the disabled attribute?

staubina replied:
I will take a look at this.

cahrens commented on common/test/acceptance/tests/lms/test_problem_types.py:
```diff
@@ -286,9 +286,6 @@ def test_problem_type_a11y(self):
-                'aria-roles',  # TODO: AC-491
                 'checkboxgroup',  # TODO: AC-491
                 'radiogroup',  # TODO: AC-491
                 'section',  # TODO: AC-491
```
Created https://openedx.atlassian.net/browse/TNL-5574 for the "section" ignore.

cahrens replied:
Created https://openedx.atlassian.net/browse/TNL-5574 for "label" ignore.

cahrens commented:
@bjacobel Can you please review the SASS in this PR? All of our feature stories have landed now.

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -401,10 +394,7 @@ div.problem {
-        width: 25px;
-        height: 20px;
-        background: url('#{$static-path}/images/partially-correct-icon.png') center center no-repeat;
+        @include status-icon($partiallycorrect, $asterisk-icon);
```
should be consistent about snake casing, in other places this appears as `partially-correct`

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -923,47 +901,62 @@ div.problem {
-  .show {
+  .problem-action-button-wrapper {
+    display: inline-block;
+    @include border-right(1px solid $light-gray1);
```
@Includes go at the top of their block (variables, includes, extends, properties)

https://github.com/edx/ux-pattern-library/wiki/Styleguide:-Sass-&-CSS#property-and-declaration-order

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -923,47 +901,62 @@ div.problem {
+
+  .problem-action-btn {
+    @include margin-right($baseline / 5);
+    @media print {
```
Usually @media blocks go at the bottom of a selector

(it's not specifically stated in our styleguide but you can see it in [some examples](https://github.com/edx/ux-pattern-library/wiki/Styleguide:-Sass-&-CSS#sass-specifically))

staubina commented:
I tested the RTL And i18n and everything looked good. 👍

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -996,69 +989,142 @@ div.problem {
-      font-style: normal;
-      font-size: 0.9em;
-    }
+    > p {
```
I would state this as `.dedicated-solution > p {` and remove this level of nesting, as `dedicated-solution` is given no rules of its own - I think it's cleaner and gives context to what `>` means

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -996,69 +989,142 @@ div.problem {
+  div.capa_alert {
+    margin-top: $baseline;
+    padding: 8px 12px;
+    border: 1px solid #ebe8bf;
```
Use a variable for this color

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -996,69 +989,142 @@ div.problem {
+    padding: 8px 12px;
+    border: 1px solid #ebe8bf;
+    border-radius: 3px;
+    background: #fffcdd;
```
ditto

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -996,69 +989,142 @@ div.problem {
+
+  .notification {
+    margin-top: $baseline / 2;
+    padding: ($baseline * 0.4) ($baseline / 2) ($baseline / 5);
```
RTL, also this padding rule only has three arguments - it has to be either 1, 2, or 4

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -996,69 +989,142 @@ div.problem {
+    margin-top: $baseline / 2;
+    padding: ($baseline * 0.4) ($baseline / 2) ($baseline / 5);
+    line-height: 1.5em;
+    min-width: auto;
```
I don't think `min-width: auto` does anything?

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -996,69 +989,142 @@ div.problem {
+    &.success {
+      @include notification-by-type($success-color);
+    }
+    &.error {
```
Spacing after each block/selector

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -996,69 +989,142 @@ div.problem {
+    }
+    .icon {
+      position: relative;
+      @include float(left);
```
@includes to the top

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -996,69 +989,142 @@ div.problem {
+      background-image: none;
+      box-shadow: none;
+    }
+    &:focus {
```
Every selector like this should have an empty line in between

bjacobel commented on common/lib/xmodule/xmodule/css/capa/display.scss:
```diff
@@ -996,69 +989,142 @@ div.problem {
-    font-size: 0.9em;
+  .review-btn {
+    color: $blue;
+    &.sr {
```
Why are we setting the color of a screenreader-only element?

bjacobel commented on common/static/sass/edx-pattern-library-shims/base/_variables.scss:
```diff
@@ -0,0 +1,221 @@
+    @include margin-right(3 * $baseline/ 4);
+    color: $color;
+  }
+}
```
Missing newline

bjacobel commented on lms/static/sass/_build-lms-v1.scss:
```diff
@@ -90,3 +90,6 @@
 @import 'shame';     // used for any bad-form/orphaned scss
+
+// CAPA Problem Feedback
+@import 'edx-pattern-library-shims/buttons';
```
This had a trailing newline, but it was removed

bjacobel commented on lms/static/sass/partials/base/_variables.scss:
```diff
@@ -211,14 +211,13 @@ $shadow-d1: rgba(0,0,0,0.4) !default;
-$error-color: rgb(253, 87, 87) !default;
-$warning-color: rgb(181,42,103) !default;
+$error-color: rgb(203, 7, 18) !default;
+$warning-color: rgb(255, 192 ,31) !default;
```
Comma should be shifted

bjacobel replied:
Comma still needs to be shifted

cptvitamin commented:
There is a current test failure related to ```<section> ```  elements that do not have headings as their first child. This is a best practice (but not a hard failure).  I have been testing our current implementation of the section element that wraps each question and think we could solve this issue and improve the user experience.  Notes are here: https://openedx.atlassian.net/browse/TNL-5574?focusedCommentId=217790&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-217790

cahrens commented:
@cptvitamin Which test failure are you referring to (this PR currently has passing tests)? I believe that @muzaffaryousaf will be addressing TNL-5574.

cahrens commented:
Bok choy test failure is a known flaky. I will update the sandbox to pick up these changes.

alisan617 commented:
jenkins run bokchoy

alisan617 commented:
@bjacobel I have addressed your comments. @cahrens There are also some alignment fix for multiple choice/check all that apply label in Studio and the pixel shift in the correctness icon above Submit button you mentioned in TNL-5543

alisan617 commented:
jenkins run js

alisan617 commented:
jenkins run a11y

cahrens commented:
@alisan617 The review button is not rendering properly on Safari.

![image](https://cloud.githubusercontent.com/assets/484484/18843842/b45c44d8-83e8-11e6-97db-2bb950738b93.png)

cahrens commented:
FYI, I am looking into the a11y failures caused by my last commit.

muzaffaryousaf commented on lms/templates/problem.html:
```diff
@@ -5,38 +5,106 @@
 
 <%namespace name='static' file='static_content.html'/>
-<h3 class="hd hd-2 problem-header">
+<h3 class="hd hd-2 problem-header" id="${ id }-problem-title" aria-describedby="${ id }-problem-progress" tabindex="-1">
```
@cahrens Any idea why we're using `id` here and what if we use `short_id` aka `problem_id` instead of this. Actually i want to refer this heading to the question title and unfortunately we don't have this id at that point, responsetype.py but we do have short_id that is being used to all other places as an identifier.

cahrens replied:
@muzaffaryousaf I don't think that any of us were aware of short_id. What is the difference between the 2? Are you asking for us to change all id usages to short_id?

FYI @staubina

cahrens replied:
Could we just pass short_id to the template as the id variable instead of changing templates?

muzaffaryousaf replied:
@cahrens `id` is the location of any block i.e. `block-v1:edX+DemoX+Demo_Course+type@problem+block@02ba1809d8584577aa703d7714e6a588-problem-title`  while `short_id` is `02ba1809d8584577aa703d7714e6a588` from this example.

cahrens replied:
Thank you, @muzaffaryousaf. Given how pervasive the usage of id is, I'm leaning towards not changing all id usages to short_id. Would you be OK changing the specific instances that you need to be short_id in your own PR?

muzaffaryousaf replied:
We need the id of this h3 element to refer it from the [section element ](https://github.com/edx/edx-platform/pull/13550/files#diff-38722d7e328c77e747d36e85e7ab2c0fR265) for a11y purpose. We already have `short_id` at some point and then just need to pass it to the response type but for `id` we have to pass it from step 1 i.e. passing it to LonCAPAProblem and then passing it to the extract_html and then to response type but i would prefer to use short_id since we're already using it for response types and LonCAPAProblem https://github.com/edx/edx-platform/blob/master/common/lib/capa/capa/capa_problem.py#L154. You can back track this id in `capa_base.py:314`.

cahrens replied:
I think you may have misunderstood me. I'm fine with this usage being short_id, but was asking if you can change it in your PR (which I assume is off of this branch). I don't think I want to change all usages of id to short_id in this particular PR.

muzaffaryousaf replied:
Great. I've changed the id to short_id for problem_title.

cahrens commented:
@alisan617 I have noticed a vertical alignment bug related to the answer feedback that appears after an input field.

If you click Submit and then Show Answer, you see this nice alignment of the answer with the icon:

![image](https://cloud.githubusercontent.com/assets/484484/18885139/606e357c-84b8-11e6-8c96-4932c9724eb0.png)

However, if you then modify the value in the text field, which causes the icon to go away, the answer text shifts up.

![image](https://cloud.githubusercontent.com/assets/484484/18885153/70ea3dd8-84b8-11e6-9458-4de55c5be03f.png)

muhammad-ammar commented:
@cahrens Can you please rebase this branch so that i can rebase my codeinput a11y branch.

alisan617 commented:
@muhammad-ammar I rebased. @cahrens  I will fix the vertical alignment on show answer later today.

cahrens commented:
FYI, no unit test failures, coverage was aborted. Following up with @jzoldak about cause.

jzoldak commented:
jenkins run python

cahrens commented:
jenkins run bokchoy

staubina commented:
jenkins run bokchoy

cahrens commented:
jenkins run lettuce

cahrens commented:
Note: Lettuce tests all passed (https://build.testeng.edx.org/job/edx-platform-lettuce-pr/26330/) with https://github.com/edx/edx-platform/pull/13052/commits/bdafc6abdafe953bb74c87377af17233333f11c0 cherry-picked onto this branch.

I am now going to remove that commit.

cahrens commented:
@staubina @alisan617 @sstack22 @cptvitamin @chris-mike any official celebratory thumbs-ups on this PR before merging? 

We still have a few remaining issues to resolve (see below), but these have not deemed to be blocking.

Remaining issues:

1. [TNL-5644](https://openedx.atlassian.net/browse/TNL-5644), #13613
2. [TNL-5549](https://openedx.atlassian.net/browse/TNL-5549)
3. [TNL-5642](https://openedx.atlassian.net/browse/TNL-5642), #13646

staubina commented:
👍 :shipit:

alisan617 commented:
🎉  +100 days old PR + 160 comments

cahrens commented:
jenkins run bokchoy

cahrens commented:
@alisan617 @staubina we seem to have a bug on this branch that the Submit button is not disabled if the problem is closed (past due date). :( I'm not going to merge until we figure this out (and add test coverage).

staubina commented:
@cahrens That is no good, although I see that a fix may have already been applied.

cahrens commented:
@alisan617 I'm seeing the (minor) issue again that the "button bar" shifts down when you first select an option in multiple choice and radio button problems. You fixed this once before, but perhaps it crept back in with the spacing work?

chris-mike commented:
👍 👍 👍

cahrens commented on lms/templates/problem.html:
```diff
@@ -30,12 +30,12 @@ <h3 class="hd hd-2 problem-header" id="${ short_id }-problem-title" aria-describ
         % if demand_hint_possible:
         <span class="problem-action-button-wrapper">
-            <button class="hint-button problem-action-btn btn-default btn-small" data-value="${_('Hint')}" ${'' if should_enable_next_hint else 'disabled'}><span class="icon fa fa-question" aria-hidden="true"></span>${_('Hint')}</button>
+            <button type="button" class="hint-button problem-action-btn btn-default btn-small" data-value="${_('Hint')}" ${'' if should_enable_next_hint else 'disabled'}><span class="icon fa fa-question" aria-hidden="true"></span>${_('Hint')}</button>
```
This change was necessary because of #13422. FYI @muzaffaryousaf @muhammad-ammar and @cptvitamin

cahrens commented:
jenkins run python

cahrens commented:
jenkins run python

staubina commented:
jenkins run python

alisan617 commented:
@cahrens can I get a hot-fix for (minor) "button bar" shifts down now? It is one line of css.

cahrens commented:
I'm going to go ahead and merge this-- these Python unit tests are failing sporadically on master as well (and @benpatterson has been made aware of them). They did pass a couple of runs ago on this branch, but then I hit a different Python sporadic issue:

https://build.testeng.edx.org/job/edx-platform-python-unittests-pr/26413/

@alisan617 Please create a PR against master (you can reuse the submit-capa sandbox if you update it with your branch), and I will review it in the morning. If I leave this particular PR open any longer, I may go crazy.

andy-armstrong commented:
Congratulations @cahrens, @alisan617 and @staubina!
